src/promptflow/promptflow/_sdk/_service/apis/collector.py

```diff
@@ -44,34 +46,43 @@ def trace_collector():
                     # TODO: persist with batch
                     span = Span._from_protobuf_object(span, resource=resource)
                     span._persist()
-                    _try_write_trace_to_cosmosdb(span)
+                    all_spans.append(span)
+
+        # Create a new thread to write trace to cosmosdb to avoid blocking the main thread
+        ThreadWithContextVars(target=_try_write_trace_to_cosmosdb, args=(all_spans,)).start()
```

Comment:
do we need to implement some persisted queues to avoid data loss?

Reply:
Good point! Maybe we could improve this later and make sure cosmosDB operation won't affect local to local trace in recent release by current code?

Reply:
Offline synced, will fix this later.